contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -5375,6 +5373,264 @@ get_partition_function(int16 dbid, const char *partition_scheme_name)
+	dsc = RelationGetDescr(rel);
+
+	/* Build a tuple to insert. */
+	new_record[Anum_bbf_partition_depend_dbid - 1] = Int16GetDatum(dbid);
+	new_record[Anum_bbf_partition_depend_scheme_name - 1] = CStringGetTextDatum(partition_scheme_name);
+	new_record[Anum_bbf_partition_depend_table_schema_name - 1] = CStringGetTextDatum(schema_name);
+	new_record[Anum_bbf_partition_depend_table_name - 1] = CStringGetTextDatum(table_name);
+
+	tuple = heap_form_tuple(dsc, new_record, new_record_nulls);
+
+	/* Insert new record in the table. */
+	CatalogTupleInsert(rel, tuple);
+
+	heap_freetuple(tuple);
+	table_close(rel, RowExclusiveLock);
+}
+
+/*
+ * remove_entry_from_bbf_partition_depend
+ *	Removes an entry from the sys.babelfish_partition_depend catalog.
+ */
+void
+remove_entry_from_bbf_partition_depend(int16 dbid, char *schema_name, char *table_name)
+{
+	Relation	rel;
+	HeapTuple	tuple;
+	ScanKeyData	scanKey[3];
+	SysScanDesc	scan;
+
+	rel = table_open(get_bbf_partition_depend_oid(), RowExclusiveLock);
+	
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_partition_depend_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_partition_depend_table_schema_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(schema_name));
+
+	ScanKeyEntryInitialize(&scanKey[2], 0, 
+				Anum_bbf_partition_depend_table_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(table_name));
+
+	scan = systable_beginscan(rel, get_bbf_partition_depend_idx_oid(),
+					false, NULL, 3, scanKey);
+	
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		CatalogTupleDelete(rel, &tuple->t_self);
+	}
+
+	systable_endscan(scan);
+	table_close(rel, RowExclusiveLock);
+}
+
+/*
+ * rename_table_update_bbf_partition_depend_catalog
+ * 	Updates table_name in sys.babelfish_partition_depend catalog for
+ * 	RENAME TABLE command to have consistency with the new names.
+ */
+void
+rename_table_update_bbf_partition_depend_catalog(RenameStmt *stmt)
+{
+	Relation	rel;
+	HeapTuple	tuple, new_tuple;
+	TupleDesc	dsc;
+	ScanKeyData	scanKey[3];
+	SysScanDesc	scan;
+	Datum		new_record[BBF_PARTITION_DEPEND_NUM_COLS];
+	bool		new_record_nulls[BBF_PARTITION_DEPEND_NUM_COLS];
+	bool		new_record_replace[BBF_PARTITION_DEPEND_NUM_COLS];
+	char		*logical_schema_name;
+	char		*table_name = stmt->relation->relname;
+	int16		dbid;
+
+	/* Find the logical schema name and dbid from physical schema name. */
+	logical_schema_name = (char *) get_logical_schema_name(stmt->relation->schemaname, true);
+
+	if (!logical_schema_name) /* not a tsql schema */
+		return;
+
+	dbid = get_dbid_from_physical_schema_name(stmt->relation->schemaname, false);
+
+	if (!is_bbf_partitioned_table(dbid, logical_schema_name, table_name))
```

Comment:
Inside is_bbf_partitioned_table we are looking up in bbf_partition_depend and then later in this function we are again looking up this table. Kindly change to lookup this table only once.